perun-core/src/main/java/cz/metacentrum/perun/core/impl/Utils.java

```diff
@@ -2021,4 +2049,279 @@ public static MapSqlParameterSource getMapSqlParameterSourceToSearchUsersOrMembe
+	 */
+	private static void sendIdentityAddedAlertToPreferredMail(PerunSession sess, UserExtSource newUes, String userEmail) {
+		sendIdentityAddedAlert(sess, newUes, userEmail,
+				UES_ADDED_PREFERRED_MAIL, DEFAULT_IDENTITY_ADDED_PREF_MAIL,
+				UES_ADDED_PREFERRED_MAIL_SUBJECT, DEFAULT_IDENTITY_ADDED_PREF_MAIL_SUBJECT);
+	}
+
+	/**
+	 * Sends alert to the email of the new given userExtSource.
+	 *
+	 * @param sess session
+	 * @param newUes new userExtSource
+	 * @param uesMail email of the newUserExtSource
+	 */
+	private static void sendIdentityAddedAlertToNewIdentityMail(PerunSession sess, UserExtSource newUes, String uesMail) {
+		sendIdentityAddedAlert(sess, newUes, uesMail,
+				UES_ADDED_UES_MAIL, DEFAULT_IDENTITY_ADDED_UES_MAIL,
+				UES_ADDED_UES_MAIL_SUBJECT, DEFAULT_IDENTITY_ADDED_UES_MAIL_SUBJECT);
+	}
+
+	/**
+	 * Sends alert about the given ues being removed.
+	 *
+	 * @param sess session
+	 * @param removedUes the removed ues
+	 * @param uesAttrs attributes of the removed ues
+	 * @param email email, where the alert is send
+	 * @param templateName name of the template for the message
+	 * @param defaultTemplate default template for the message
+	 * @param subjectName name of the subject template
+	 * @param defaultSubject default template for the subject
+	 */
+	private static void sendIdentityRemovedAlert(PerunSession sess, UserExtSource removedUes, List<Attribute> uesAttrs,
+	                                             String email, String templateName, String defaultTemplate,
+	                                             String subjectName, String defaultSubject) {
+		String message = getAlertTemplate(sess, templateName, defaultTemplate);
+		message = message.replaceAll(LOGIN_PLACEHOLDER, removedUes.getLogin());
+		message = insertOrganization(message, uesAttrs);
+		message = insertCurrentTime(message);
+
+		String subject = getAlertTemplate(sess, subjectName, defaultSubject);
+
+		Utils.sendEmail(subject, message, email);
+	}
+
+	/**
+	 * Sends alert about the given ues being added.
+	 *
+	 * @param sess session
+	 * @param newUes the added ues
+	 * @param email email, where the alert is send
+	 * @param templateName name of the template for the message
+	 * @param defaultTemplate default template for the message
+	 * @param subjectName name of the subject template
+	 * @param defaultSubject default template for the subject
+	 */
+	private static void sendIdentityAddedAlert(PerunSession sess, UserExtSource newUes, String email, String templateName,
+	                                           String defaultTemplate, String subjectName, String defaultSubject) {
+		String message = getAlertTemplate(sess, templateName, defaultTemplate);
+		message = message.replaceAll(LOGIN_PLACEHOLDER, newUes.getLogin());
+		message = insertOrganization(sess, message, newUes);
+		message = insertCurrentTime(message);
+
+		String subject = getAlertTemplate(sess, subjectName, defaultSubject);
+
+		Utils.sendEmail(subject, message, email);
```

Comment:
Whole action of adding/removing identity can now fail if mail can't be sent. I'm not sure if it`s desired. We should discuss it with requestor and probably internally and unify this behavior on other actions too.

Reply:
Good point, however, I would suggest just catching such exceptions and logging them. I don't think we would like the whole process to fail when some email couldn't be sent.

However, I will ask the requestor, if such behavior is OK.